sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveUnion.scala

```diff
@@ -182,6 +182,15 @@ object ResolveUnion extends Rule[LogicalPlan] {
             // union have consistent schema.
             aliased += foundAttr
             Alias(addFields(foundAttr, target), foundAttr.name)()
+          case (source: StructType, target: StructType)
+            if !allowMissingCol && !source.sameType(target) &&
+              target.toAttributes.map(attr => attr.name).sorted
+                == source.toAttributes.map(x => x.name).sorted =>
```

Comment:
This looks a bit fragile as it doesn't consider case sensitivity.

Reply:
UnionByName with allowMissing columns true add it as the missing column in case of case senistive  attributes for both the scenarios  spark.sql.caseSensitive as true and false

```
case class UnionClass2(a: Int, c: String)
case class UnionClass4(A: Int, b: Long)
case class UnionClass1a(a: Int, b: Long, nested: UnionClass2)
case class UnionClass1c(a: Int, b: Long, nested: UnionClass4)


 val df1 = Seq((0, UnionClass1a(0, 1L, UnionClass2(1, "2")))).toDF("id", "a")
 val df2 = Seq((1, UnionClass1c(1, 2L, UnionClass4(2, 3L)))).toDF("id", "a")

case 1 - set spark.sql.caseSensitive=false

scala> spark.sql("set spark.sql.caseSensitive=false")
res6: org.apache.spark.sql.DataFrame = [key: string, value: string]

scala> var unionDf = df1.unionByName(df2, true)
unionDf: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [id: int, a: struct<a: int, b: bigint ... 1 more field>]

scala> unionDf.schema.toDDL
res7: String = `id` INT,`a` STRUCT<`a`: INT, `b`: BIGINT, `nested`: STRUCT<`a`: INT, `b`: BIGINT, `c`: STRING, `A`: INT>>

case 2 -> when spark.sql.caseSensitive is enabled 
scala> spark.sql("set spark.sql.caseSensitive=true")
res2: org.apache.spark.sql.DataFrame = [key: string, value: string]


scala> 

scala> var unionDf = df1.unionByName(df2, true)
unionDf: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [id: int, a: struct<a: int, b: bigint ... 1 more field>]

scala> unionDf.schema.toDDL
res3: String = `id` INT,`a` STRUCT<`a`: INT, `b`: BIGINT, `nested`: STRUCT<`A`: INT, `a`: INT, `b`: BIGINT, `c`: STRING>>
```

for UnionbyName without allowMissing -> we cannot add the missing column ,it should give the exception with schema not same so union cannot be done  so that is the reason for not converting into lower case and than do the comparison

```
> var unionDf = df1.unionByName(df2)
org.apache.spark.sql.AnalysisException: Union can only be performed on tables with the compatible column types. struct<a:int,b:bigint,nested:struct<A:int,b:bigint>> <> struct<a:int,b:bigint,nested:struct<a:int,c:string>> at the second column of the second table;
```

Reply:
I think he means more like

```
df1 = spark.createDataFrame([Row(nested=Row(a=1, b=2))])
df2 = spark.createDataFrame([Row(nested=Row(B=1, A=2))])
df1.unionByName(df2)
```

These wouldn't get merged without doing case insensitive comparisons

Reply:
This validation is not required 
```target.toAttributes.map(attr => attr.name).sorted == source.toAttributes.map(x => x.name).sorted ```
Since the we need to sort both left and right side in case of struct and sortStructFields method recursively sorts in all the nested struct.

Now after removing this validation case sensitive attributes with different sequence are also working 
as its working for the case sensitive attributes with same sequence